scan-payment/src/main/java/com/getbouncer/scan/payment/ml/ExpiryDetect.kt

```diff
@@ -0,0 +1,136 @@
+package com.getbouncer.scan.payment.ml
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.util.Size
+import com.getbouncer.scan.framework.Loader
+import com.getbouncer.scan.framework.ResourceLoader
+import com.getbouncer.scan.framework.ml.TFLAnalyzerFactory
+import com.getbouncer.scan.framework.ml.TensorFlowLiteAnalyzer
+import com.getbouncer.scan.framework.util.indexOfMax
+import com.getbouncer.scan.payment.R
+import com.getbouncer.scan.payment.hasOpenGl31
+import com.getbouncer.scan.payment.scale
+import com.getbouncer.scan.payment.toRGBByteBuffer
+import org.tensorflow.lite.Interpreter
+import java.io.FileNotFoundException
+import java.nio.ByteBuffer
+
+private val TRAINED_IMAGE_SIZE = Size(80, 36)
+
+private const val NUM_CLASS = 11
+private const val NUM_PREDICTIONS = 17
+private const val BACKGROUND_CLASS = 10
+
+val ASPECT_RATIO = TRAINED_IMAGE_SIZE.height.toFloat() / TRAINED_IMAGE_SIZE.width.toFloat()
+
+class ExpiryDetect private constructor(interpreter: Interpreter) :
+    TensorFlowLiteAnalyzer<ExpiryDetect.Input, ByteBuffer,
+        ExpiryDetect.Prediction,
+        Array<Array<Array<FloatArray>>>>(interpreter) {
+
+    data class Input(val image: Bitmap)
+
+    data class Prediction(val expiry: Expiry?)
+
+    data class Expiry(val month: Int, val year: Int) {
+        override fun toString() = "${month.toString().padStart(2, '0')}/$year"
+    }
+
+    data class Digit(val digit: Int, val confidence: Float)
+
+    override val name: String = Factory.NAME
+
+    override fun buildEmptyMLOutput() = arrayOf(
+        arrayOf(
+            Array(17) {
+                FloatArray(NUM_CLASS)
+            }
+        )
+    )
+
+    override fun interpretMLOutput(data: Input, mlOutput: Array<Array<Array<FloatArray>>>): Prediction {
+        val output = mutableListOf<Digit>()
+        for (idx in 0 until 17) {
+            val maxIdx = mlOutput[0][0][idx].indexOfMax()!!
+            output.add(Digit(maxIdx, mlOutput[0][0][idx][maxIdx]))
+        }
+
+        val digits = greedyNonMaxSuppression(output).filter { it != 10 }
+
+        return if (digits.size == 4 || (digits.size == 5 && digits[2] == 1)) {
```

Comment:
Some of my cards have a full date, i.e. MM/DD/YY as the expiration. Should we take that into account here as well?

Reply:
hmm, not sure whether the current text detector will recognize those, as they're trained on MM/YY format. can you send me some images? i'll test and see if the model detects them at all